### PR TITLE
Bf gismo export

### PR DESCRIPTION
**Pull request by FSchwar**
# Overview
Fixed issues when exporting multipatches with `None` as field values. 

## Addressed issues
* multipatches with `None` in fields raised several errors

## Checklists
* [ ] Documentations are up-to-date.
* [ ] Added example(s)
* [ ] Added test(s)

### CODE REVIEW

File: splinepy/io/gismo.py [hunk 1]
```diff
@@ -302,13 +309,45 @@ def export(
     patch_range = _ET.SubElement(
         multipatch_element, "patches", type="id_range"
     )
-    patch_range.text = (
-        f"{index_offset} " f"{len(multipatch.patches) - 1 + index_offset}"
-    )
+    # patch_range.text = (
+    #     f"{index_offset} " f"{len(multipatch.patches) - 1 + index_offset}"
+    # )
```
Comment (jzwar) [inline splinepy/io/gismo.py:314]: Deprecated

File: splinepy/io/gismo.py [hunk 2]
```diff
@@ -109,10 +109,14 @@ def _array_to_text(array, is_matrix, as_b64):
             [" ".join([str(xx) for xx in x]) for x in supports]
         )
 
+    id_count = -1
```
Comment (jzwar) [inline splinepy/io/gismo.py:112]: Please add a comment why this is done

File: splinepy/io/gismo.py [hunk 3]
```diff
@@ -302,13 +309,45 @@ def export(
...
+        with open(fname + ".fields.xml", "wb") as f:
+            f.write(file_content)
+
+    n_patches = _spline_to_ET(
+        xml_data,
+        multipatch,
+        index_offset,
+        as_base64=as_base64,
+    )
+    assert n_patches + 1 == len(multipatch.patches)
```
Comment (jzwar) [inline splinepy/io/gismo.py:346]: if raise

**Review by jzwar — COMMENTED** [review]
Left some remarks,

why do the tests fail. Please check!

**Comment by FSchwar** [comment]
`test_gismo` fails in `test_gismo_export`at line 84 - where the lines of the `base_file` and the temporary exported file are compared.
I manually compared the files and it seems that only the order of the lines is off.

To be more precise, the definition of the `boundaryConditions` in the `base_file` is before the definition of `Geometry`, whereas in the temporary file (newly exported one), the definition of the boundary conditions is at the bottom of the xml file.

File: splinepy/io/gismo.py [hunk 4]
```diff
@@ -302,13 +311,45 @@ def export(
...
+        field_xml.find("MultiPatch").find("patches").text = (
+            f"{index_offset} " f"{n_patches + index_offset}"
+        )
+        if int(_python_version.split(".")[1]) >= 9 and indent:
+            _ET.indent(field_xml)
+        file_content = _ET.tostring(field_xml)
+        with open(fname + ".fields.xml", "wb") as f:
+            f.write(file_content)
+
+    n_patches = _spline_to_ET(
```
Comment (jzwar) [inline splinepy/io/gismo.py:339]: This Block of Code must stay where it was otherwise the order will change

File: splinepy/io/gismo.py [hunk 5]
```diff
@@ -80,7 +80,7 @@ def _array_to_text(array, is_matrix, as_b64):
             field_mask = _np.arange(len(multipatch.fields))
         supports = _np.array(
             [
-                (i, j, 0)
+                (i, k, 0)
```
Comment (jzwar) [inline splinepy/io/gismo.py:83]: I think we don't need the enumerate block here do we? J is unused

File: splinepy/io/gismo.py (hunk 4 above)
Comment (FSchwar) [reply splinepy/io/gismo.py]: It's successful now, thanks!

File: splinepy/io/gismo.py [hunk 6]
```diff
@@ -109,10 +109,16 @@ def _array_to_text(array, is_matrix, as_b64):
...
+    # set to -1 since loop starts with id_count += 1, therefore the first
+    # patch starts with id = 0
+    id_count = -1
     for id, spline in enumerate(multipatch.patches):
         if fields_only:
             # Check supports
             support = supports[supports[:, 0] == id, 1]
+            if support.size == 0:
+                continue
+            id_count += 1
```
[the remark this answers is not in the document]
Comment (FSchwar) [reply splinepy/io/gismo.py:121]: not needed (as discussed)

File: splinepy/io/gismo.py (hunk 5 above)
Comment (FSchwar) [reply splinepy/io/gismo.py]: was changed back again in a later commit for successful field mask

**Comment by jzwar** [comment]
There is still one issue that is unsolved in opinion at the moment. I think the reader will break if the spline ids are not contiguous, meaning if we only use splines `[0,1,2,4,5]` it will most likely break. This is because the the `support` matrix is referring to the splines in their global configuration, whereas the splines themselves are simply counted. There is two possible solutions to this issue. Either we introduce another column in the support matrix, where the global ids are matched a local id, this would have to be updated [here](https://github.com/tataratat/pygadjoints/blob/34aa2f287a0b8b8e453a865548d112bb80128b75/src/pygadjoints/py_elasticity.hpp#L610C17-L610C24) and in the respective diffusion field. Second would not be a fix, but rather a warning, where the user is warned (in case a `id_count!=id`. 

Please implement the latter, and I will make sure to update whenever this becomes an issue. Just make an elaborate warning, whenever this problem pops up

**Pull request merged**